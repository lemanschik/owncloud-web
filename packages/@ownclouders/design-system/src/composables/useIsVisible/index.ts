import { onBeforeUnmount, ref, watch } from 'vue'

/**
 *
 * @param {Ref<Element>} target - ref with element to be observed
 * @param {('show'|'showHide')} mode - showHide shows and hides the element on screen enter or leave, show only detects entering the screen and the keeps it rendered
 * @param {string} rootMargin - margin that will be added around the element to detect visibility
 * @returns {{isVisible: Ref<boolean>}}
 */
export const useIsVisible = ({ target, mode = 'show', rootMargin = '100px' }) => {
  const isSupported = window && 'IntersectionObserver' in window
  if (!isSupported) {
    return {
      isVisible: ref(true)
    }
  }

  const isVisible = ref(false)
  const observer = new IntersectionObserver(
    (intersectionObserverEntries: IntersectionObserverEntry[]) => {
      /**
       * In some edge cases intersectionObserverEntries contains 2 entries with the first one having wrong rootBounds.
       * This happens for some reason when the table is being re-sorted immediately after being rendered.
       * Therefore we always check the last entry for isIntersecting.
       */
      const isIntersecting = intersectionObserverEntries.at(-1).isIntersecting

      isVisible.value = isIntersecting
      /**
       * if given mode is `showHide` we need to keep the observation alive.
       */
      if (mode === 'showHide') {
        return
      }
      /**
       * if the mode is `show` which is the default, the implementation needs to unsubscribe the target from the observer
       */
      if (!isVisible.value) {
        return
      }

      observer.unobserve(target.value)
    },
    {
      rootMargin
    }
  )

  watch(target, () => {
    observer.observe(target.value)
  })

  onBeforeUnmount(() => observer.disconnect())

  return {
    isVisible
  }
}
