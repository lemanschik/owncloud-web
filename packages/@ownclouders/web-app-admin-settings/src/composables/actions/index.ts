export * from './general'
export * from './groups'
export * from './users'
