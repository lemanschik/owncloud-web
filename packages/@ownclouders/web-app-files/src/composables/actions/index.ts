export * from './files'
export * from './helpers'
export * from './spaces'
