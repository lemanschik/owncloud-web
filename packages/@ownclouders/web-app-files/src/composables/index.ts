export * from './actions'
export * from './archiverService'
export * from './pagination'
export * from './resourcesViewDefaults'
export * from './router'
export * from './selection'
export * from './shares'
export * from './spaces'
export * from './viewMode'
