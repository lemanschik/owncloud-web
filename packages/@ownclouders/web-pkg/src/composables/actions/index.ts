export * from './helpers'
export * from './files'
export * from './spaces'
export * from './types'

export * from './useActionsShowDetails'
