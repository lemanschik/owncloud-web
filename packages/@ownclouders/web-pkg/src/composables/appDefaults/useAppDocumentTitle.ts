import { Ref, computed, unref } from 'vue'
import { basename } from 'path'
import { FileContext } from './types'
import { useAppMeta } from './useAppMeta'
import { useDocumentTitle } from './useDocumentTitle'
import { Store } from 'vuex'
import { RouteLocationNormalizedLoaded } from 'vue-router'
import { MaybeRef } from 'web-pkg'
import { useGettext } from 'vue3-gettext'

interface AppDocumentTitleOptions {
  store: Store<any>
  applicationId: string
  applicationName?: MaybeRef<string>
  currentFileContext: Ref<FileContext>
  currentRoute?: Ref<RouteLocationNormalizedLoaded>
}

export function useAppDocumentTitle({
  store,
  applicationId,
  applicationName,
  currentFileContext,
  currentRoute
}: AppDocumentTitleOptions): void {
  const appMeta = useAppMeta({ applicationId, store })
  const { $gettext } = useGettext()

  const titleSegments = computed(() => {
    const baseTitle =
      basename(unref(unref(currentFileContext)?.fileName)) ||
      $gettext((unref(currentRoute)?.meta?.title as string) || '')
    const meta = unref(unref(appMeta).applicationMeta)

    return [baseTitle, unref(applicationName) || meta.name || meta.id].filter(Boolean)
  })

  useDocumentTitle({
    titleSegments
  })
}
