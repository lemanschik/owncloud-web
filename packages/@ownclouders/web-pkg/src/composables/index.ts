export * from './ability'
export * from './appDefaults'
export * from './authContext'
export * from './capability'
export * from './clientService'
export * from './configuration'
export * from './driveResolver'
export * from './loadingService'
export * from './localStorage'
export * from './router'
export * from './previewService'
export * from './spaces'
export * from './sort'
export * from './store'
export * from './fileListHeaderPosition'
