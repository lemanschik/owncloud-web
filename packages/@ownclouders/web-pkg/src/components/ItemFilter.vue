<template>
  <div class="item-filter oc-flex" :class="`item-filter-${filterName}`">
    <oc-filter-chip
      :filter-label="filterLabel"
      :selected-item-names="selectedItems.map((i) => i[displayNameAttribute])"
      @clear-filter="clearFilter"
      @show-drop="showDrop"
    >
      <template #default>
        <oc-text-input
          v-if="showOptionFilter && filterableAttributes.length"
          ref="filterInputRef"
          v-model="filterTerm"
          class="item-filter-input oc-mb-m oc-mt-s"
          autocomplete="off"
          :label="optionFilterLabel === '' ? $gettext('Filter list') : optionFilterLabel"
        />
        <div ref="itemFilterListRef">
          <oc-list class="item-filter-list">
            <li v-for="(item, index) in displayedItems" :key="index" class="oc-my-xs">
              <component
                :is="isSelectionAllowed(item) ? 'oc-button' : 'div'"
                class="item-filter-list-item oc-flex oc-flex-left oc-flex-middle oc-width-1-1 oc-p-xs"
                justify-content="left"
                appearance="raw"
                :data-test-value="item[displayNameAttribute]"
                @click="toggleItemSelection(item)"
              >
                <div>
                  <oc-checkbox
                    size="large"
                    class="item-filter-checkbox"
                    :label="$gettext('Toggle selection')"
                    :model-value="isItemSelected(item)"
                    :disabled="!isSelectionAllowed(item)"
                    hide-label
                    @update:model-value="toggleItemSelection(item)"
                    @click.stop
                  />
                </div>
                <div>
                  <slot name="image" :item="item" />
                </div>
                <div>
                  <slot name="item" :item="item" />
                </div>
              </component>
            </li>
          </oc-list>
        </div>
      </template>
    </oc-filter-chip>
  </div>
</template>

<script lang="ts">
import { defineComponent, nextTick, onMounted, ref, unref, watch } from 'vue'
import Fuse from 'fuse.js'
import Mark from 'mark.js'
import omit from 'lodash-es/omit'
import { defaultFuseOptions, useRoute, useRouteQuery, useRouter } from 'web-pkg'
import { queryItemAsString } from 'web-pkg/src/composables/appDefaults'

export default defineComponent({
  name: 'ItemFilter',
  props: {
    filterLabel: {
      type: String,
      required: true
    },
    filterName: {
      type: String,
      required: true
    },
    optionFilterLabel: {
      type: String,
      required: false,
      default: ''
    },
    showOptionFilter: {
      type: Boolean,
      required: false,
      default: false
    },
    items: {
      type: Array,
      required: true
    },
    allowMultiple: {
      type: Boolean,
      required: false,
      default: false
    },
    displayNameAttribute: {
      type: String,
      required: false,
      default: 'name'
    },
    filterableAttributes: {
      type: Array,
      required: false,
      default: () => []
    }
  },
  emits: ['selectionChange'],
  setup: function (props, { emit }) {
    const router = useRouter()
    const currentRoute = useRoute()
    const filterInputRef = ref()
    const selectedItems = ref([])
    const displayedItems = ref(props.items)
    const markInstance = ref(null)
    const itemFilterListRef = ref(null)

    const queryParam = `q_${props.filterName}`
    const currentRouteQuery = useRouteQuery(queryParam)
    const setRouteQuery = () => {
      router.push({
        query: {
          ...omit(unref(currentRoute).query, [queryParam]),
          ...(!!unref(selectedItems).length && {
            [queryParam]: unref(selectedItems)
              .reduce((acc, item) => {
                acc += `${item.id}+`
                return acc
              }, '')
              .slice(0, -1)
          })
        }
      })
    }

    const isItemSelected = (item) => {
      return !!unref(selectedItems).find((s) => s.id === item.id)
    }
    const isSelectionAllowed = (item) => {
      return props.allowMultiple || !unref(selectedItems).length || isItemSelected(item)
    }
    const toggleItemSelection = (item) => {
      if (!isSelectionAllowed(item)) {
        return
      }
      if (isItemSelected(item)) {
        selectedItems.value = unref(selectedItems).filter((s) => s.id !== item.id)
      } else {
        selectedItems.value.push(item)
      }
      emit('selectionChange', unref(selectedItems))
      setRouteQuery()
    }

    const sortItems = (items) => {
      const selectedItemIds = unref(selectedItems).map((i) => i.id)
      return items.sort(
        (a: any, b: any) =>
          (selectedItemIds.includes(b.id) as any) - (selectedItemIds.includes(a.id) as any) ||
          a[props.displayNameAttribute].localeCompare(b[props.displayNameAttribute])
      )
    }

    const filterTerm = ref()
    const filter = (items, filterTerm) => {
      if (!(filterTerm || '').trim()) {
        return items
      }
      const usersSearchEngine = new Fuse(items, {
        ...defaultFuseOptions,
        keys: props.filterableAttributes as any
      })

      return usersSearchEngine.search(filterTerm).map((r) => r.item)
    }
    const clearFilter = () => {
      selectedItems.value = []
      emit('selectionChange', unref(selectedItems))
      setRouteQuery()
    }

    const setDisplayedItems = (items) => {
      displayedItems.value = sortItems(items)
    }

    const showDrop = async () => {
      setDisplayedItems(props.items)
      await nextTick()
      unref(filterInputRef).focus()
    }

    watch(filterTerm, () => {
      setDisplayedItems(filter(props.items, unref(filterTerm)))
      if (unref(itemFilterListRef)) {
        markInstance.value = new Mark(unref(itemFilterListRef))
        unref(markInstance).unmark()
        unref(markInstance).mark(unref(filterTerm), {
          element: 'span',
          className: 'highlight-mark'
        })
      }
    })

    onMounted(() => {
      const queryStr = queryItemAsString(unref(currentRouteQuery))
      if (queryStr) {
        const ids = queryStr.split('+')
        selectedItems.value = props.items.filter((s: any) => ids.includes(s.id))
      }
    })

    return {
      clearFilter,
      displayedItems,
      filterInputRef,
      filterTerm,
      isItemSelected,
      isSelectionAllowed,
      itemFilterListRef,
      queryParam,
      selectedItems,
      setDisplayedItems,
      showDrop,
      toggleItemSelection
    }
  }
})
</script>

<style lang="scss">
.item-filter {
  &-list {
    li {
      &:first-child {
        margin-top: 0 !important;
      }
      &:last-child {
        margin-bottom: 0 !important;
      }
    }

    &-item {
      line-height: 1.5;
      gap: 8px;

      &:hover {
        background-color: var(--oc-color-background-hover) !important;
      }
    }
  }
}
</style>

