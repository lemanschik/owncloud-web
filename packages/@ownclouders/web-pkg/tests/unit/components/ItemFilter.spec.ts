import ItemFilter from 'web-pkg/src/components/ItemFilter.vue'
import { defaultComponentMocks, defaultPlugins, mount } from 'web-test-helpers'
import { queryItemAsString } from 'web-pkg/src/composables/appDefaults'

jest.mock('web-pkg/src/composables/appDefaults')

const filterItems = [
  { id: '1', name: 'Albert Einstein' },
  { id: '2', name: 'Marie Curie' }
]

const selectors = {
  filterInput: '.item-filter-input input',
  checkboxStub: 'oc-checkbox-stub',
  filterListItem: '.item-filter-list-item',
  clearBtn: '.oc-filter-chip-clear'
}

describe('ItemFilter', () => {
  it('renders all items', () => {
    const { wrapper } = getWrapper()
    expect(wrapper.html()).toMatchSnapshot()
  })
  it('can use a custom attribute as display name', () => {
    const filterItems = [
      { id: '1', displayName: 'Albert Einstein' },
      { id: '2', displayName: 'Marie Curie' }
    ]
    const { wrapper } = getWrapper({
      props: { displayNameAttribute: 'displayName', items: filterItems }
    })
    expect(wrapper.html()).toMatchSnapshot()
  })
  describe('filter', () => {
    it('renders the input field when enabled', () => {
      const { wrapper } = getWrapper({
        props: { showOptionFilter: true, filterableAttributes: ['name'] }
      })
      expect(wrapper.find(selectors.filterInput).exists()).toBeTruthy()
    })
    it.each([
      { filterTerm: '', expectedResult: 2 },
      { filterTerm: 'Albert', expectedResult: 1 },
      { filterTerm: 'invalid', expectedResult: 0 }
    ])('filters on input', async (data) => {
      const { wrapper } = getWrapper({
        props: { showOptionFilter: true, filterableAttributes: ['name'] }
      })
      await wrapper.find(selectors.filterInput).setValue(data.filterTerm)
      expect(wrapper.findAll(selectors.filterListItem).length).toBe(data.expectedResult)
    })
  })
  describe('selection', () => {
    it('allows selection of multiple items on click', async () => {
      const { wrapper } = getWrapper({ props: { allowMultiple: true } })
      expect(wrapper.emitted('selectionChange')).toBeFalsy()
      let selectionChangeEmits = 0
      for (const item of wrapper.findAll(selectors.filterListItem)) {
        expect(item.findComponent<any>(selectors.checkboxStub).props('modelValue')).toBeFalsy()
        await item.trigger('click')
        selectionChangeEmits += 1
        expect(wrapper.emitted('selectionChange').length).toBe(selectionChangeEmits)
        expect(item.findComponent<any>(selectors.checkboxStub).props('modelValue')).toBeTruthy()
      }
      expect(wrapper.vm.selectedItems.length).toBe(wrapper.findAll(selectors.filterListItem).length)
    })
    it('does not allow selection of multiple items on click when disabled', async () => {
      const { wrapper } = getWrapper({ props: { allowMultiple: false } })
      const first = wrapper.findAll(selectors.filterListItem).at(0)
      await first.trigger('click')
      expect(wrapper.emitted('selectionChange').length).toBe(1)
      expect(first.findComponent<any>(selectors.checkboxStub).props('modelValue')).toBeTruthy()

      const second = wrapper.findAll(selectors.filterListItem).at(1)
      await second.trigger('click')
      expect(wrapper.emitted('selectionChange').length).toBe(1)
      expect(second.findComponent<any>(selectors.checkboxStub).props('modelValue')).toBeFalsy()
      expect(second.findComponent<any>(selectors.checkboxStub).props('disabled')).toBeTruthy()
      expect(wrapper.vm.selectedItems.length).toBe(1)
    })
    it('can always de-select the current selected item', async () => {
      const { wrapper } = getWrapper({ props: { allowMultiple: false } })
      const item = wrapper.findAll(selectors.filterListItem).at(0)
      await item.trigger('click')
      await item.trigger('click')
      expect(wrapper.emitted('selectionChange').length).toBe(2)
      expect(item.findComponent<any>(selectors.checkboxStub).props('modelValue')).toBeFalsy()
      expect(wrapper.vm.selectedItems.length).toBe(0)
    })
    it('clears the selection when the clear-button is being clicked', async () => {
      const { wrapper } = getWrapper({ props: { allowMultiple: false } })
      const item = wrapper.findAll(selectors.filterListItem).at(0)
      await item.trigger('click')
      await wrapper.find(selectors.clearBtn).trigger('click')
      expect(wrapper.emitted('selectionChange').length).toBe(2)
      expect(item.findComponent<any>(selectors.checkboxStub).props('modelValue')).toBeFalsy()
      expect(wrapper.vm.selectedItems.length).toBe(0)
    })
  })
  describe('route query', () => {
    it('sets the selected item as route query param', async () => {
      const { wrapper, mocks } = getWrapper()
      const item = wrapper.findAll(selectors.filterListItem).at(0)
      const currentRouteQuery = (mocks.$router.currentRoute as any).query
      expect(mocks.$router.push).not.toHaveBeenCalled()
      await item.trigger('click')
      expect(currentRouteQuery[wrapper.vm.queryParam]).toBeDefined()
      expect(mocks.$router.push).toHaveBeenCalled()
    })
    it('sets the selected items initially when given via query param', () => {
      const { wrapper } = getWrapper({ initialQuery: '1' })
      expect(wrapper.vm.selectedItems).toEqual([filterItems[0]])
    })
  })

  describe('label prop', () => {
    it('sets the correct label using getLabel computed property', () => {
      const label = 'Filter groups'
      const { wrapper } = getWrapper({
        props: {
          showOptionFilter: true,
          filterableAttributes: ['name'],
          optionFilterLabel: label
        }
      })
      expect(wrapper.find('.item-filter-input label').text()).toBe(label)
    })

    it('sets the default label using getLabel computed property when no prop is set', () => {
      const label = undefined
      const { wrapper } = getWrapper({
        props: {
          showOptionFilter: true,
          filterableAttributes: ['name'],
          optionFilterLabel: label
        }
      })
      expect(wrapper.find('.item-filter-input label').text()).toBe('Filter list')
    })
  })
})

function getWrapper({ props = {}, initialQuery = '' }: any = {}) {
  jest.mocked(queryItemAsString).mockImplementation(() => initialQuery)
  const mocks = defaultComponentMocks()
  return {
    mocks,
    wrapper: mount(ItemFilter, {
      props: {
        filterLabel: 'Users',
        filterName: 'users',
        items: filterItems,
        ...props
      },
      slots: {
        item(data: any) {
          return props.displayNameAttribute ? data.item[props.displayNameAttribute] : data.item.name
        }
      },
      global: {
        plugins: [...defaultPlugins()],
        mocks,
        stubs: { OcCheckbox: true }
      }
    })
  }
}
