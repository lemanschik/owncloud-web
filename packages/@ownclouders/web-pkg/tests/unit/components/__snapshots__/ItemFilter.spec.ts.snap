// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ItemFilter can use a custom attribute as display name 1`] = `
<div class="item-filter oc-flex item-filter-users">
  <div class="oc-filter-chip oc-flex">
    <button class="oc-button oc-rounded oc-button-m oc-button-justify-content-center oc-button-gap-m oc-button-passive oc-button-passive-raw oc-filter-chip-button oc-pill" id="oc-filter-chip-5" type="button">
      <!-- @slot Content of the button -->
      <!--v-if-->
      <span class="oc-text-truncate oc-filter-chip-label">Users</span>
      <!--v-if-->
      <span class="oc-icon oc-icon-s oc-icon-passive">
        <!---->
      </span>
    </button>
    <div class="oc-drop oc-box-shadow-medium oc-rounded oc-filter-chip-drop" id="oc-drop-6">
      <div class="oc-card oc-card-body oc-background-secondary oc-p-s">
        <!--v-if-->
        <div>
          <ul class="oc-list oc-my-rm oc-mx-rm item-filter-list">
            <li class="oc-my-xs">
              <button class="oc-button oc-rounded oc-button-m oc-button-justify-content-left oc-button-gap-m oc-button-passive oc-button-passive-raw item-filter-list-item oc-flex oc-flex-left oc-flex-middle oc-width-1-1 oc-p-xs" data-test-value="Albert Einstein" type="button">
                <!-- @slot Content of the button -->
                <div>
                  <oc-checkbox-stub class="item-filter-checkbox" disabled="false" hidelabel="true" id="oc-checkbox-7" label="Toggle selection" modelvalue="false" outline="false" size="large"></oc-checkbox-stub>
                </div>
                <div></div>
                <div>Albert Einstein</div>
              </button>
            </li>
            <li class="oc-my-xs">
              <button class="oc-button oc-rounded oc-button-m oc-button-justify-content-left oc-button-gap-m oc-button-passive oc-button-passive-raw item-filter-list-item oc-flex oc-flex-left oc-flex-middle oc-width-1-1 oc-p-xs" data-test-value="Marie Curie" type="button">
                <!-- @slot Content of the button -->
                <div>
                  <oc-checkbox-stub class="item-filter-checkbox" disabled="false" hidelabel="true" id="oc-checkbox-8" label="Toggle selection" modelvalue="false" outline="false" size="large"></oc-checkbox-stub>
                </div>
                <div></div>
                <div>Marie Curie</div>
              </button>
            </li>
          </ul>
        </div>
      </div>
    </div>
    <!--v-if-->
  </div>
</div>
`;

exports[`ItemFilter renders all items 1`] = `
<div class="item-filter oc-flex item-filter-users">
  <div class="oc-filter-chip oc-flex">
    <button class="oc-button oc-rounded oc-button-m oc-button-justify-content-center oc-button-gap-m oc-button-passive oc-button-passive-raw oc-filter-chip-button oc-pill" id="oc-filter-chip-1" type="button">
      <!-- @slot Content of the button -->
      <!--v-if-->
      <span class="oc-text-truncate oc-filter-chip-label">Users</span>
      <!--v-if-->
      <span class="oc-icon oc-icon-s oc-icon-passive">
        <!---->
      </span>
    </button>
    <div class="oc-drop oc-box-shadow-medium oc-rounded oc-filter-chip-drop" id="oc-drop-2">
      <div class="oc-card oc-card-body oc-background-secondary oc-p-s">
        <!--v-if-->
        <div>
          <ul class="oc-list oc-my-rm oc-mx-rm item-filter-list">
            <li class="oc-my-xs">
              <button class="oc-button oc-rounded oc-button-m oc-button-justify-content-left oc-button-gap-m oc-button-passive oc-button-passive-raw item-filter-list-item oc-flex oc-flex-left oc-flex-middle oc-width-1-1 oc-p-xs" data-test-value="Albert Einstein" type="button">
                <!-- @slot Content of the button -->
                <div>
                  <oc-checkbox-stub class="item-filter-checkbox" disabled="false" hidelabel="true" id="oc-checkbox-3" label="Toggle selection" modelvalue="false" outline="false" size="large"></oc-checkbox-stub>
                </div>
                <div></div>
                <div>Albert Einstein</div>
              </button>
            </li>
            <li class="oc-my-xs">
              <button class="oc-button oc-rounded oc-button-m oc-button-justify-content-left oc-button-gap-m oc-button-passive oc-button-passive-raw item-filter-list-item oc-flex oc-flex-left oc-flex-middle oc-width-1-1 oc-p-xs" data-test-value="Marie Curie" type="button">
                <!-- @slot Content of the button -->
                <div>
                  <oc-checkbox-stub class="item-filter-checkbox" disabled="false" hidelabel="true" id="oc-checkbox-4" label="Toggle selection" modelvalue="false" outline="false" size="large"></oc-checkbox-stub>
                </div>
                <div></div>
                <div>Marie Curie</div>
              </button>
            </li>
          </ul>
        </div>
      </div>
    </div>
    <!--v-if-->
  </div>
</div>
`;
